### DISCUSSION TITLE
How to create custom commands

### DISCUSSION THREAD

**Discussion by KennethHoff** [microsoft/aspire#6015] [Q&A]
I've swapped over to using Aspire 9 Preview and in doing so I noticed that custom commands has been added. However, after looking at how the built-in commands are implemented, I noticed that they retrieve an internal interface/class in order to do their "restart services" functionality. 

What I want is a button that toggles between two different environment variables on an executable. How would I go forward with this?

**Comment by davidfowl** [comment]
> What I want is a button that toggles between two different environment variables on an executable. How would I go forward with this?

You can't change environment variables at runtime, the model doesn't allow for that currently.

**Reply by KennethHoff** [reply]
How would I go forward with simply adding a custom restart button them? Manually editing an env. file and then restarting the executable seems like a good enough hack for now

**Reply by BaconSoap** [reply]
I played around with a workaround for a similar problem (dynamic flags for a migration/general CLI tool) that feels not-too-bad, depending on your views of reflection and global state 🙈. I shared it as its own discussion #8502 and an at-own-risk [gist](https://gist.github.com/BaconSoap/0100f636ff62cec4b76f55e9f7de2c1b). I implemented the Start method, but it should be quick to copy-paste for [StopResourceAsync](https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs#L315), and from there copy the [built-in restart](https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/ApplicationModel/CommandsConfigurationExtensions.cs#L88-L95) command. You can then use the callback form of WithEnvironment() to conditionally set that flag before the app starts

Also added some thoughts on potential features to the tail end of the gist that I feel would drastically simplify/stabilize this story and open up more safe escape-hatch options for dev tooling - a minimal non-reflection-based proxy would be great, but understand not wanting to commit to an interface quite yet before more integrated options are fleshed out.

Thanks for the starting point!
